sled-agent/src/rack_setup/plan/sled.rs

```diff
@@ -53,6 +55,42 @@ pub struct Plan {
     pub config: Config,
 }
 
+impl FromStr for Plan {
+    type Err = String;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        #[derive(Deserialize)]
+        struct ShadowPlan {
+            pub rack_id: Uuid,
+            pub sleds: BTreeMap<SocketAddrV6, StartSledAgentRequest>,
+            pub config: Config,
+        }
+        #[derive(Deserialize)]
+        struct ShadowPlanV1 {
+            pub rack_id: Uuid,
+            pub sleds: BTreeMap<SocketAddrV6, StartSledAgentRequest>,
+            pub config: ConfigV1,
+        }
+        let v2_err = match serde_json::from_str::<ShadowPlan>(&value) {
+            Ok(plan) => {
+                return Ok(Plan {
+                    rack_id: plan.rack_id,
+                    sleds: plan.sleds,
+                    config: plan.config,
+                })
+            }
+            Err(e) => format!("unable to parse Plan: {e:?}"),
+        };
+        serde_json::from_str::<ShadowPlanV1>(&value)
+            .map(|v1| Plan {
+                rack_id: v1.rack_id,
+                sleds: v1.sleds,
+                config: v1.config.into(),
+            })
+            .map_err(|_| v2_err)
```

Comment:
... and looks like the same behavior here, I assume that if we added a new version you'd want us to do:

- try to parse v3
- try to parse v2
- try to parse v1
- if all those fail, return the parse error from v3

Reply:
Right.  It's the same pattern in both places, which makes turning this into a Trait potentially useful.